### PR TITLE
Collection extensions upd

### PR DESCRIPTION
**Pull request by Banalny-Banan**
`Shuffle<T>(this IEnumerable<T> enumerable, int iterations = 1)` - added documentation for thrown exceptions & added an actual shuffle instead of infinite recursive call

`AddRange<T>(this IEnumerable<T> enumerable, IEnumerable<T> collection)` - before was calling `AddItem()` in a loop. (multiple `Concat()`s internaly). It wouldnt even work because `AddItem()` is a pure method. Now using one `Concat()`.

`AddRange<T>(this T[] array, IEnumerable<T> collection)` - same changes.

`AddRange<T>(this HashSet<T> hashset, IEnumerable<T> collection)` - before was calling `Add()` in a loop. Now using `UnionWith()`.

### CODE REVIEW

**Review by ALEXWARELLC — APPROVED** [review]

File: Exiled.API/Extensions/CollectionExtensions.cs [hunk 1]
```diff
@@ -70,7 +70,16 @@ public static IEnumerable<T> Shuffle<T>(this IEnumerable<T> enumerable, int iter
                 throw new ArgumentOutOfRangeException(nameof(iterations));
 
             T[] array = enumerable.ToArray();
-            array.Shuffle(iterations);
+
+            for (int i = 0; i < iterations; i++)
+            {
+                for (int j = array.Length - 1; j > 0; j--)
```
Comment (VALERA771) [inline Exiled.API/Extensions/CollectionExtensions.cs:76]:
```suggestion
                for (int j = array.Length - 1; j >= 0; j--)
```

**Comment by ALEXWARELLC** [comment]
> Maybe we should add an "Instance" field to Plugin<> class instead of making all those GetPlugin() methods ? I remember it existing in the example plugin and assigned in OnEnabled() We could just do it in base class

Isn't this outside the scope of the PR?

File: Exiled.API/Extensions/CollectionExtensions.cs [hunk 2]
```diff
@@ -61,16 +59,27 @@ public static IEnumerable<T> RemoveSpecified<T>(this IEnumerable<T> enumerable,
         /// <typeparam name="T">The type of the elements of <see cref="IEnumerable{T}"/>.</typeparam>
         /// <param name="enumerable">The <see cref="IEnumerable{T}"/>.</param>
         /// <param name="iterations">The amount of times to repeat the shuffle operation.</param>
+        /// <exception cref="ArgumentNullException">Thrown when the <see cref="IEnumerable{T}"/> is null.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown when the amount of iterations is less than 1.</exception>
         /// <returns>A shuffled version of the <see cref="IEnumerable{T}"/>.</returns>
-        public static IEnumerable<T> Shuffle<T>(this IEnumerable<T> enumerable, int iterations = 1)
```
Comment (VALERA771) [inline Exiled.API/Extensions/CollectionExtensions.cs:65]: Please, leave here `IEnumerable<T>`

File: Exiled.API/Extensions/CollectionExtensions.cs [hunk 3]
```diff
@@ -70,7 +70,16 @@ public static IEnumerable<T> Shuffle<T>(this IEnumerable<T> enumerable, int iter
                 throw new ArgumentOutOfRangeException(nameof(iterations));
 
             T[] array = enumerable.ToArray();
-            array.Shuffle(iterations);
```
[the remark this answers is not in the document]
Comment (Misfiy) [reply Exiled.API/Extensions/CollectionExtensions.cs:73]: I believe that would call the Array Shuffle not the exiled extension, could be wrong

[the remark this answers is not in the document]
Comment (VALERA771) [reply Exiled.API/Extensions/CollectionExtensions.cs:73]: There is a `CollectionExtensions::ShuffleList` method compitable with arrays

**Review by VALERA771 — CHANGES_REQUESTED** [review]
Use a Nw extension (ShuffleList)

File: Exiled.API/Extensions/CollectionExtensions.cs (hunk 2 above)
Comment (ALEXWARELLC) [reply Exiled.API/Extensions/CollectionExtensions.cs:65]: What's the issue here?

File: Exiled.API/Extensions/CollectionExtensions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ALEXWARELLC) [reply Exiled.API/Extensions/CollectionExtensions.cs:73]: > There is a `CollectionExtensions::ShuffleList` method compitable with arrays

I don't think it's possible to use.

**Review by ALEXWARELLC — APPROVED** [review]
I'm ok with this

**Comment by VALERA771** [comment]
Okay. Now you need to target dev branch (read exiled announcements if any questions)

**Review by VALERA771 — APPROVED** [review]

**Review by ALEXWARELLC — APPROVED** [review]

**Review by Misfiy — APPROVED** [review]

**Pull request merged**